pip/pip-278.md

```diff
@@ -0,0 +1,128 @@
+
+
+# High Level Design
+
+To make the whole topic compaction service pluggable, we need to abstract `TopicCompactionService` interface, it can provide the capability that the compactor has and provide the read API to read entries from compacted data.
+
+We should combine `CompactedTopicImpl` and `TwoPhaseCompactor` to pulsar implementation of topic compaction service and make behavior with the current implementation consistent.
+
+
+# Detailed Design
+
+## Design & Implementation Details
+
+* Define a standard TopicCompactionService interface.
+
+  ```java
+  public interface TopicCompactionService {
+      /**
+       * Compact the topic.
+       * Topic Compaction is a key-based retention mechanism. it will keep the most recent value for a given key and
+       * user will read compacted data from TopicCompactionService.
+       *
+       * @return a future that will be completed when the compaction is done.
+       */
+      CompletableFuture<Void> compact();
+
+      /**
+       * Read the compacted entries from the topic.
+       *
+       * @param startPosition         the position to start reading from.
+       * @param numberOfEntriesToRead the number of entries to read.
+       * @return a future that will be completed with the list of entries, this list can is null.
+       */
+      CompletableFuture<List<Entry>> readCompactedEntries(PositionImpl startPosition, int numberOfEntriesToRead);
+
+      /**
+       * Read the last compacted entry from the TopicCompactionService.
+       *
+       * @return a future that will be completed with the compacted last entry, this entry can is null.
+       */
+      CompletableFuture<Entry> readCompactedLastEntry();
+
+      /**
+       * Get the last compacted position from the TopicCompactionService.
+       *
+       * @return a future that will be completed with the last compacted position, this position can is null.
+       */
+      CompletableFuture<PositionImpl> getCompactedLastPosition();
+  }
+  ```
+
+* Define a standard CompactionServiceFactory interface to manage `TopicCompactionService`.
+
+  ```java
+  public interface CompactionServiceFactory extends AutoCloseable {
+  
+      CompletableFuture<Void> initialize(PulsarService pulsarService);
+  
+      CompletableFuture<TopicCompactionService> newTopicCompactionService(String topic);
+  }
+  ```
+
+* Implement `PulsarCompactionServiceFactory` and `PulsarCompactionService`
+
+* Combining `CompactedTopicImpl` and `TwoPhaseCompactor` to `PulsarTopicCompactionService`
+
+* Rename `CompactorSubscription` to `PulsarCompactorSubscription`, since it is only applicable to the implementation of Pulsar.
+
+* For CompactorMetrics keep the current implementation and it only support `PulsarTopicCompactionService` currently , in the future will use the `Otel` API or other mertics API  instead, and customize compactedService should implement it.
```

Comment:
Hi，@asafm 
I remember you have been working on the metrics lately. What do you think of the place?  thanks! :)